src/charm.py

```diff
@@ -2,77 +2,71 @@
 
+"""Charmed Machine Operator for Apache ZooKeeper."""
+
 import logging
-import subprocess
 
+from charms.kafka.v0.kafka_snap import KafkaSnap
 from ops.charm import CharmBase
 from ops.main import main
-from ops.model import ActiveStatus, BlockedStatus, MaintenanceStatus
-
-from kafka_helpers import install_packages, merge_config
+from ops.model import Relation
 
 logger = logging.getLogger(__name__)
 
 
-class KafkaCharm(CharmBase):
+class ZooKeeperCharm(CharmBase):
+    """Charmed Operator for ZooKeeper."""
+
     def __init__(self, *args):
         super().__init__(*args)
-        self.name = "kafka"
+        self.name = "zookeeper"
+        self.snap = KafkaSnap()
 
         self.framework.observe(getattr(self.on, "install"), self._on_install)
         self.framework.observe(
             getattr(self.on, "cluster_relation_joined"), self._on_cluster_relation_joined
         )
         self.framework.observe(getattr(self.on, "leader_elected"), self._on_leader_elected)
         self.framework.observe(
-            getattr(self.on, "get_server_properties_action"), self._on_get_server_properties_action
+            getattr(self.on, "get_zookeeper_properties_action"), self._on_get_properties_action
         )
+        self.framework.observe(
+            getattr(self.on, "get_snap_apps_action"), self._on_get_snap_apps_action
+        )
+        self.framework.observe(getattr(self.on, "config_changed"), self._on_config_changed)
 
     @property
-    def _relation(self):
+    def _relation(self) -> Relation:
         return self.model.get_relation("cluster")
 
     def _on_install(self, _) -> None:
-        try:
-            self.unit.status = MaintenanceStatus("installing packages")
-            install_packages()
-            self.unit.status = ActiveStatus()
-        except:
-            self.unit.status = BlockedStatus("failed to install packages")
-
-    def _on_leader_elected(self, _):
-        return
+        """Handler for on_install event."""
+        self.unit.status = self.snap.install_kafka_snap()
 
-    def _on_cluster_relation_joined(self, _):
+    def _on_leader_elected(self, _) -> None:
         return
 
-    def _on_config_changed(self, _) -> None:
-        self._start_services()
-
-        if not isinstance(self.unit.status, BlockedStatus):
-            self.unit.status = ActiveStatus()
-
-    def _start_services(self) -> None:
+    def _on_cluster_relation_joined(self, _) -> None:
         return
 
-    def _run_command(self, cmd: list) -> bool:
-        proc = subprocess.Popen(cmd)
-        for line in iter(getattr(proc.stdout, "readline"), ""):
-            logger.debug(line)
-        proc.wait()
-        return proc.returncode == 0
-
-    def _on_get_server_properties_action(self, event):
-        """Handler for users to copy currently active config for passing to `juju config`"""
+    def _on_config_changed(self, _) -> None:
+        """Handler for config_changed event."""
+        self._start_service()
```

Comment:
Wouldn't be better to call this in `on_start` to avoid calling it multiple times if some config changes? I don't know if it's intended to have configs in the charm in the future and/or if the `snap start` command is idempotent.

Reply:
You're right, good catch.

`snap start` won't do anything if the service is already running, but for a `config_changed` event it should be `snap restart`. Thanks!